index.js

```diff
@@ -2,104 +2,42 @@
-  let addons = addon.project.addonPackages;
-
-  for (let i = 0; i < requiredDataPackages.length; i++) {
-    let requiredPackageName = requiredDataPackages[i];
-    if (addons[requiredPackageName] === undefined) {
-      throw new Error(
-        `ember-m3 ${pkg.version} requires the peerDependency ${requiredPackageName} to be installed.`
-      );
-    }
-  }
-
-  let storeVersion = addons['@ember-data/store'].pkg.version;
-
-  if (semver.lt(storeVersion, minVersionForUsingOwnPackages)) {
-    throw new Error(
-      `To use your own @ember-data/<pkg> versions with ember-m3 their versions must be at least ${minVersionForUsingOwnPackages}. Found @ember-data/store with version ${storeVersion}.`
-    );
-  }
-
-  for (let i = 0; i < allDataPackages.length; i++) {
-    let addonName = allDataPackages[i];
-    let addon = addons[addonName];
-
-    if (addon && addonName !== 'ember-inflector') {
-      let pkgVersion = addons[addonName].pkg.version;
-      if (pkgVersion !== storeVersion) {
-        throw new Error(
-          `All @ember-data/<pkg> packages must have matching versions. Found ${addonName} ${pkgVersion} which does not match @ember-data/store's version of ${storeVersion}.`
-        );
-      }
-    }
-  }
-}
+const debug = require('debug')('ember-m3');
+const checkDependencyVersions = require('./src/dependency-checks');
 
 module.exports = {
   name: 'ember-m3',
 
   init() {
-    let ret = this._super.init.call(this, ...arguments);
-
-    if (!hasDataPackage(this) && hasOwnDataPackages(this)) {
-      assertOwnDataPackagesValid(this);
-    }
-
-    return ret;
+    this._hasInitializedDependencyChecks = false;
+    return this._super.init.apply(this, arguments);
   },
 
   included() {
-    this._super.included.call(this, ...arguments);
+    this._super.included.apply(this, arguments);
     this.configureBabelOptions();
   },
 
   shouldIncludeChildAddon(addon) {
-    if (!hasDataPackage(this) && !hasOwnDataPackages(this)) {
+    if (!this._hasInitializedDependencyChecks) {
+      // we can't initialize in init because we need to super first
+      // but super calls shouldIncludeChildAddon which needs this
+      // so we initialize from the first call to shouldIncludeChildAddon
+      this._hasInitializedDependencyChecks = true;
+      this._excludeChildDataAddons = checkDependencyVersions(this.project);
+    }
+
+    if (this._excludeChildDataAddons !== true) {
+      debug(`\t✅  including dependency ${addon.name} because including all children`);
       return true;
     }
     if (addon.name.startsWith('@ember-data') || addon.name === 'ember-inflector') {
-      /*
-      console.log(
-        `⚠️  ember-m3 is excluding ${addon.name} version ${addon.pkg.version} from the build`
+      debug(
+        `\t⚠️  ember-m3 is excluding its own copy of ${addon.name} version ${addon.pkg.version} from the build`
       );
-      */
       return false;
     }
+    debug(`\t✅  including dependency ${addon.name}`);
```

Comment:
Does this run when you build apps?

Reply:
yes

Reply:
apps won't see these logs though unless they run with `DEBUG=*` or `DEBUG=ember-m3`